# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/webrtc/signaling/gtest/MockCall.h
# Commit: 877a0fcc6da9
# Full Hash: 877a0fcc6da9fde4a783cfe30d0aa0b260427ac3
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 9
# Description:
#   Bug 1654112 - Merge conduit controllers into conduits, refactor away setters, and make unittests async. r=bwc,ng
#   
#   The main point of this patch is to harden how conduits handle configuration
#   changes. With conduit controller classes acting in a layer on top of conduits
#   (and their old sync style APIs), for instance
# ==============================================================================

diff -r 970ed33359b5 -r 877a0fcc6da9 media/webrtc/signaling/gtest/MockCall.h
--- a/media/webrtc/signaling/gtest/MockCall.h	Tue Aug 10 14:13:49 2021 +0200
+++ b/media/webrtc/signaling/gtest/MockCall.h	Mon Aug 16 15:57:10 2021 +0200
@@ -7,6 +7,8 @@
 
 #include "gmock/gmock.h"
 #include "mozilla/Assertions.h"
+#include "mozilla/Maybe.h"
+#include "mozilla/media/MediaUtils.h"
 #include "WebrtcCallWrapper.h"
 #include "PeerConnectionCtx.h"
 
@@ -15,16 +17,16 @@
 #include "call/call.h"
 
 namespace test {
+class MockCallWrapper;
 
 class MockAudioSendStream : public webrtc::AudioSendStream {
  public:
-  MockAudioSendStream() : mConfig(nullptr) {}
+  explicit MockAudioSendStream(RefPtr<MockCallWrapper> aCallWrapper)
+      : mCallWrapper(std::move(aCallWrapper)) {}
 
-  const webrtc::AudioSendStream::Config& GetConfig() const override {
-    return mConfig;
-  }
+  const webrtc::AudioSendStream::Config& GetConfig() const override;
 
-  void Reconfigure(const Config& config) override { mConfig = config; }
+  void Reconfigure(const Config& config) override;
 
   void Start() override {}
 
@@ -46,12 +48,15 @@
 
   virtual ~MockAudioSendStream() {}
 
-  webrtc::AudioSendStream::Config mConfig;
+  const RefPtr<MockCallWrapper> mCallWrapper;
   webrtc::AudioSendStream::Stats mStats;
 };
 
 class MockAudioReceiveStream : public webrtc::AudioReceiveStream {
  public:
+  explicit MockAudioReceiveStream(RefPtr<MockCallWrapper> aCallWrapper)
+      : mCallWrapper(std::move(aCallWrapper)) {}
+
   void Start() override {}
 
   void Stop() override {}
@@ -68,20 +73,21 @@
     return mRtpSources;
   }
 
-  void Reconfigure(const Config& config) override {}
+  void Reconfigure(const Config& config) override;
   bool SetBaseMinimumPlayoutDelayMs(int delay_ms) override { return false; }
   int GetBaseMinimumPlayoutDelayMs() const override { return 0; }
 
   virtual ~MockAudioReceiveStream() {}
 
+  const RefPtr<MockCallWrapper> mCallWrapper;
   webrtc::AudioReceiveStream::Stats mStats;
   std::vector<webrtc::RtpSource> mRtpSources;
 };
 
 class MockVideoSendStream : public webrtc::VideoSendStream {
  public:
-  explicit MockVideoSendStream(webrtc::VideoEncoderConfig&& config)
-      : mEncoderConfig(std::move(config)) {}
+  explicit MockVideoSendStream(RefPtr<MockCallWrapper> aCallWrapper)
+      : mCallWrapper(std::move(aCallWrapper)) {}
 
   void Start() override {}
 
@@ -91,9 +97,7 @@
       rtc::VideoSourceInterface<webrtc::VideoFrame>* source,
       const webrtc::DegradationPreference& degradation_preference) override {}
 
-  void ReconfigureVideoEncoder(webrtc::VideoEncoderConfig config) override {
-    mEncoderConfig = config.Copy();
-  }
+  void ReconfigureVideoEncoder(webrtc::VideoEncoderConfig config) override;
 
   Stats GetStats() override { return mStats; }
 
@@ -110,12 +114,15 @@
 
   virtual ~MockVideoSendStream() {}
 
-  webrtc::VideoEncoderConfig mEncoderConfig;
+  const RefPtr<MockCallWrapper> mCallWrapper;
   webrtc::VideoSendStream::Stats mStats;
 };
 
 class MockVideoReceiveStream : public webrtc::VideoReceiveStream {
  public:
+  explicit MockVideoReceiveStream(RefPtr<MockCallWrapper> aCallWrapper)
+      : mCallWrapper(std::move(aCallWrapper)) {}
+
   void Start() override {}
 
   void Stop() override {}
@@ -150,60 +157,65 @@
 
   virtual ~MockVideoReceiveStream() {}
 
+  const RefPtr<MockCallWrapper> mCallWrapper;
   webrtc::VideoReceiveStream::Stats mStats;
 };
 
 class MockCall : public webrtc::Call {
  public:
-  MockCall()
-      : mAudioSendConfig(nullptr),
-        mVideoReceiveConfig(nullptr),
-        mVideoSendConfig(nullptr),
-        mCurrentVideoSendStream(nullptr) {}
+  explicit MockCall(RefPtr<MockCallWrapper> aCallWrapper)
+      : mCallWrapper(std::move(aCallWrapper)) {}
 
   webrtc::AudioSendStream* CreateAudioSendStream(
       const webrtc::AudioSendStream::Config& config) override {
-    mAudioSendConfig = config;
-    return new MockAudioSendStream;
+    MOZ_RELEASE_ASSERT(!mAudioSendConfig);
+    mAudioSendConfig = mozilla::Some(config);
+    return new MockAudioSendStream(mCallWrapper);
   }
 
   void DestroyAudioSendStream(webrtc::AudioSendStream* send_stream) override {
+    mAudioSendConfig = mozilla::Nothing();
     delete static_cast<MockAudioSendStream*>(send_stream);
   }
 
   webrtc::AudioReceiveStream* CreateAudioReceiveStream(
       const webrtc::AudioReceiveStream::Config& config) override {
-    mAudioReceiveConfig = config;
-    return new MockAudioReceiveStream;
+    MOZ_RELEASE_ASSERT(!mAudioReceiveConfig);
+    mAudioReceiveConfig = mozilla::Some(config);
+    return new MockAudioReceiveStream(mCallWrapper);
   }
   void DestroyAudioReceiveStream(
       webrtc::AudioReceiveStream* receive_stream) override {
+    mAudioReceiveConfig = mozilla::Nothing();
     delete static_cast<MockAudioReceiveStream*>(receive_stream);
   }
 
   webrtc::VideoSendStream* CreateVideoSendStream(
       webrtc::VideoSendStream::Config config,
       webrtc::VideoEncoderConfig encoder_config) override {
-    MOZ_RELEASE_ASSERT(!mCurrentVideoSendStream);
-    mVideoSendConfig = config.Copy();
-    mCurrentVideoSendStream = new MockVideoSendStream(encoder_config.Copy());
-    return mCurrentVideoSendStream;
+    MOZ_RELEASE_ASSERT(!mVideoSendConfig);
+    MOZ_RELEASE_ASSERT(!mVideoSendEncoderConfig);
+    mVideoSendConfig = mozilla::Some(std::move(config));
+    mVideoSendEncoderConfig = mozilla::Some(std::move(encoder_config));
+    return new MockVideoSendStream(mCallWrapper);
   }
 
   void DestroyVideoSendStream(webrtc::VideoSendStream* send_stream) override {
-    MOZ_RELEASE_ASSERT(mCurrentVideoSendStream == send_stream);
-    mCurrentVideoSendStream = nullptr;
+    mVideoSendConfig = mozilla::Nothing();
+    mVideoSendEncoderConfig = mozilla::Nothing();
     delete static_cast<MockVideoSendStream*>(send_stream);
   }
 
   webrtc::VideoReceiveStream* CreateVideoReceiveStream(
       webrtc::VideoReceiveStream::Config configuration) override {
-    mVideoReceiveConfig = configuration.Copy();
-    return new MockVideoReceiveStream;
+    MOZ_RELEASE_ASSERT(!mVideoReceiveConfig);
+    mVideoReceiveConfig = mozilla::Some(std::move(configuration));
+    return new MockVideoReceiveStream(mCallWrapper);
   }
 
   void DestroyVideoReceiveStream(
       webrtc::VideoReceiveStream* receive_stream) override {
+    mVideoReceiveConfig = mozilla::Nothing();
     delete static_cast<MockVideoReceiveStream*>(receive_stream);
   }
 
@@ -239,20 +251,19 @@
       const webrtc::BitrateSettings& preferences) override {}
 
   std::vector<webrtc::VideoStream> CreateEncoderStreams(int width, int height) {
-    const webrtc::VideoEncoderConfig& config =
-        mCurrentVideoSendStream->mEncoderConfig;
-    return config.video_stream_factory->CreateEncoderStreams(width, height,
-                                                             config);
+    return mVideoSendEncoderConfig->video_stream_factory->CreateEncoderStreams(
+        width, height, *mVideoSendEncoderConfig);
   }
 
   virtual ~MockCall(){};
 
-  webrtc::AudioReceiveStream::Config mAudioReceiveConfig;
-  webrtc::AudioSendStream::Config mAudioSendConfig;
-  webrtc::VideoReceiveStream::Config mVideoReceiveConfig;
-  webrtc::VideoSendStream::Config mVideoSendConfig;
+  const RefPtr<MockCallWrapper> mCallWrapper;
+  mozilla::Maybe<webrtc::AudioReceiveStream::Config> mAudioReceiveConfig;
+  mozilla::Maybe<webrtc::AudioSendStream::Config> mAudioSendConfig;
+  mozilla::Maybe<webrtc::VideoReceiveStream::Config> mVideoReceiveConfig;
+  mozilla::Maybe<webrtc::VideoSendStream::Config> mVideoSendConfig;
+  mozilla::Maybe<webrtc::VideoEncoderConfig> mVideoSendEncoderConfig;
   webrtc::Call::Stats mStats;
-  MockVideoSendStream* mCurrentVideoSendStream;
 };
 
 class MockCallWrapper : public mozilla::WebrtcCallWrapper {
@@ -278,7 +289,7 @@
     auto wrapper = mozilla::MakeRefPtr<MockCallWrapper>(
         state, nullptr, nullptr, nullptr,
         mozilla::dom::RTCStatsTimestampMaker(), nullptr);
-    wrapper->SetCall(mozilla::WrapUnique(new MockCall));
+    wrapper->SetCall(mozilla::WrapUnique(new MockCall(wrapper)));
     return wrapper;
   }
 